{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd2db94f-a10b-4da7-9f70-1dc623630579",
   "metadata": {},
   "source": [
    "# disease_run\n",
    "\n",
    "对disease数据集运行`run_cross_species_models`\n",
    "\n",
    "\n",
    "```shell\n",
    "conda activate\n",
    "cd ~/link/res_publish/run\n",
    "jupyter nbconvert disease_run.ipynb --to python\n",
    "\n",
    "conda activate publish\n",
    "nohup python disease_run.py > nohup_disease_run.txt &\n",
    "jobs\n",
    "\n",
    "conda activate\n",
    "echo 'finish'\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4988b915-bdf0-4ff9-9be5-e800196a876e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "p_root = Path('~/link/res_publish').expanduser()\n",
    "None if str(p_root) in sys.path else sys.path.append(str(p_root))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4bc19322-0d07-4fa6-b692-97e2abc89980",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "2024-05-06 17:32:37.269702: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-------------------------help-------------------------\n",
      "> parameter\n",
      "    p_root\t[name] res_publish\n",
      "        p_run, p_plot, p_res, p_cache, p_pdf\n",
      "    p_df_varmap\n",
      "    map_sp_reverse\n",
      "    rng\n",
      "> run\n",
      "    run_cross_species_models\n",
      "    h5ad_to_mtx\n",
      "    load_adata\n",
      "    get_path_varmap\n",
      "    find_path_from_para\n",
      "    load_normalized_adata\n",
      "\n",
      "> res\n",
      "    get_test_result_df\n",
      "    get_res_obs\n",
      "    get_adata_umap\n",
      "    show_umap\n",
      "\n",
      "> plot\n",
      "    get_color_map\n",
      "    show_color_map\n",
      "    show_color\n",
      "    plot_umap\n",
      "    savefig\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from func import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16b881e2-f011-4b07-bcf9-b4756a568ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>sp_ref</th>\n",
       "      <th>path_ref</th>\n",
       "      <th>name_ref</th>\n",
       "      <th>sp_simple_ref</th>\n",
       "      <th>sp_que</th>\n",
       "      <th>path_que</th>\n",
       "      <th>name_que</th>\n",
       "      <th>sp_simple_que</th>\n",
       "      <th>key_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChDCs</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmDCs</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChNeu</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmNeu</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChall</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmall</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChMono</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmMono</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>RAhfib</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>RAmfib</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-h-macr</td>\n",
       "      <td>h</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-ma-macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge-ma-macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_T</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_T</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_imm</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_imm</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LN</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_T</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_T</td>\n",
       "      <td>h</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LN</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_imm</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_imm</td>\n",
       "      <td>h</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_B</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_B</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue sp_ref                                           path_ref  \\\n",
       "0     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "1     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "2     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "3     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "0     RA  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "0    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "1    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "2    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "3    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "0     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "1     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "2     LN  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "3     LN  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "4     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "             name_ref sp_simple_ref   sp_que  \\\n",
       "0              LChDCs             h    mouse   \n",
       "1              LChNeu             h    mouse   \n",
       "2              LChall             h    mouse   \n",
       "3             LChMono             h    mouse   \n",
       "0              RAhfib             h    mouse   \n",
       "0          Mtb-m-macr             m    human   \n",
       "1          Mtb-m-macr             m  macaque   \n",
       "2          Mtb-m-macr             m  macaque   \n",
       "3          Mtb-m-macr             m  macaque   \n",
       "0             SLE_h_T             h    mouse   \n",
       "1           SLE_h_imm             h    mouse   \n",
       "2    merge_m_kidney_T             m    human   \n",
       "3  merge_m_kidney_imm             m    human   \n",
       "4             SLE_h_B             h    mouse   \n",
       "\n",
       "                                            path_que            name_que  \\\n",
       "0  /public/workspace/licanchengup/link/res_publis...              LCmDCs   \n",
       "1  /public/workspace/licanchengup/link/res_publis...              LCmNeu   \n",
       "2  /public/workspace/licanchengup/link/res_publis...              LCmall   \n",
       "3  /public/workspace/licanchengup/link/res_publis...             LCmMono   \n",
       "0  /public/workspace/licanchengup/link/res_publis...              RAmfib   \n",
       "0  /public/workspace/licanchengup/link/res_publis...          Mtb-h-macr   \n",
       "1  /public/workspace/licanchengup/link/res_publis...         Mtb-ma-macr   \n",
       "2  /public/workspace/licanchengup/link/res_publis...       COVID-ma-macr   \n",
       "3  /public/workspace/licanchengup/link/res_publis...       merge-ma-macr   \n",
       "0  /public/workspace/licanchengup/link/res_publis...    merge_m_kidney_T   \n",
       "1  /public/workspace/licanchengup/link/res_publis...  merge_m_kidney_imm   \n",
       "2  /public/workspace/licanchengup/link/res_publis...             SLE_h_T   \n",
       "3  /public/workspace/licanchengup/link/res_publis...           SLE_h_imm   \n",
       "4  /public/workspace/licanchengup/link/res_publis...    merge_m_kidney_B   \n",
       "\n",
       "  sp_simple_que   key_cell_type  \n",
       "0             m   sub_cell_type  \n",
       "1             m   sub_cell_type  \n",
       "2             m       cell_type  \n",
       "3             m   sub_cell_type  \n",
       "0             m   sub_cell_type  \n",
       "0             h  macr_cell_type  \n",
       "1            ma  macr_cell_type  \n",
       "2            ma  macr_cell_type  \n",
       "3            ma  macr_cell_type  \n",
       "0             m   sub_cell_type  \n",
       "1             m       cell_type  \n",
       "2             h   sub_cell_type  \n",
       "3             h       cell_type  \n",
       "4             m   sub_cell_type  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "tissue                                                          LN\n",
       "sp_ref                                                       human\n",
       "path_ref         /public/workspace/licanchengup/link/res_publis...\n",
       "name_ref                                                   SLE_h_B\n",
       "sp_simple_ref                                                    h\n",
       "sp_que                                                       mouse\n",
       "path_que         /public/workspace/licanchengup/link/res_publis...\n",
       "name_que                                          merge_m_kidney_B\n",
       "sp_simple_que                                                    m\n",
       "key_cell_type                                        sub_cell_type\n",
       "Name: 4, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_disease = 'LC,RA,Mtb,LN'.split(',')\n",
    "\n",
    "df_para = pd.concat([pd.read_csv(p_cache.joinpath(\n",
    "    'parameter_{}.csv'.format(i)\n",
    ")) for i in q_disease])\n",
    "df_para['path_ref'] = df_para['path_ref'].apply(lambda x:p_cache.joinpath(x))\n",
    "df_para['path_que'] = df_para['path_que'].apply(lambda x:p_cache.joinpath(x))\n",
    "display(df_para)\n",
    "\n",
    "for _i, _row in df_para.iterrows():\n",
    "    pass\n",
    "_row"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5e2b20-c197-443b-b2ba-04e897dbbccb",
   "metadata": {},
   "source": [
    "# RA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fbc132c1-5cbf-4955-bbc0-1da0d0d24ec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAhfib-map-RAmfib</td>\n",
       "      <td>1844</td>\n",
       "      <td>17</td>\n",
       "      <td>11707</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tag  cell_count_ref  _batch_ref  cell_count_que  _batch_que\n",
       "0  RAhfib-map-RAmfib            1844          17           11707           6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disease = 'RA'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = df_apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57ebd526-dfe2-4f8e-a893-0dde415ebdb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] RA\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 RA_h-corss-m;csMAHN;RAhfib-map-RAmfib;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 RA_h-corss-m;came;RAhfib-map-RAmfib;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240506-1732 RA_h-corss-m;SAMap;RAhfib-map-RAmfib;is_1v1=False\n",
      "\n",
      "[run finish] RA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 100, 100]    \n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c6d92ba-6a8e-47f8-840d-061e90ecb1d1",
   "metadata": {},
   "source": [
    "# LC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6454d5b5-502b-4a4a-a9c6-fb397aea422a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LChDCs-map-LCmDCs</td>\n",
       "      <td>2010</td>\n",
       "      <td>18</td>\n",
       "      <td>646</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LChNeu-map-LCmNeu</td>\n",
       "      <td>2728</td>\n",
       "      <td>18</td>\n",
       "      <td>3593</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LChall-map-LCmall</td>\n",
       "      <td>34558</td>\n",
       "      <td>18</td>\n",
       "      <td>9201</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LChMono-map-LCmMono</td>\n",
       "      <td>1176</td>\n",
       "      <td>18</td>\n",
       "      <td>771</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   tag  cell_count_ref  _batch_ref  cell_count_que  _batch_que\n",
       "0    LChDCs-map-LCmDCs            2010          18             646           9\n",
       "1    LChNeu-map-LCmNeu            2728          18            3593           9\n",
       "2    LChall-map-LCmall           34558          18            9201           9\n",
       "3  LChMono-map-LCmMono            1176          18             771           9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disease = 'LC'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = df_apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61e5543d-6fa6-43dd-986a-6e7d2de0f1a4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] LC\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;csMAHN;LChDCs-map-LCmDCs;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;came;LChDCs-map-LCmDCs;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;csMAHN;LChNeu-map-LCmNeu;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;came;LChNeu-map-LCmNeu;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;csMAHN;LChall-map-LCmall;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;came;LChall-map-LCmall;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;csMAHN;LChMono-map-LCmMono;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;came;LChMono-map-LCmMono;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;csMAHN;LChall-map-LCmall;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240506-1732 LC_h-corss-m;came;LChall-map-LCmall;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240506-1732 LC_h-corss-m;SAMap;LChDCs-map-LCmDCs;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240506-1732 LC_h-corss-m;SAMap;LChNeu-map-LCmNeu;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240506-1732 LC_h-corss-m;SAMap;LChall-map-LCmall;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240506-1732 LC_h-corss-m;SAMap;LChMono-map-LCmMono;is_1v1=False\n",
      "\n",
      "[run finish] LC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "n_epochs = [100, 100, 100]\n",
    "for _i, _row in df_para_disease.query(\"key_cell_type == 'cell_type'\").iterrows():\n",
    "    \n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5b1c5a7-abd4-4fe4-a7f7-cc8f3d9c9a34",
   "metadata": {},
   "source": [
    "# Mtb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "55575306-3cb5-4d85-b88b-8e2b2a3c21fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mtb-m-macr-map-Mtb-h-macr</td>\n",
       "      <td>15720</td>\n",
       "      <td>7</td>\n",
       "      <td>10357</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mtb-m-macr-map-Mtb-ma-macr</td>\n",
       "      <td>15720</td>\n",
       "      <td>7</td>\n",
       "      <td>18217</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mtb-m-macr-map-COVID-ma-macr</td>\n",
       "      <td>15720</td>\n",
       "      <td>7</td>\n",
       "      <td>1630</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mtb-m-macr-map-merge-ma-macr</td>\n",
       "      <td>15720</td>\n",
       "      <td>7</td>\n",
       "      <td>19847</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            tag  cell_count_ref  _batch_ref  cell_count_que  \\\n",
       "0     Mtb-m-macr-map-Mtb-h-macr           15720           7           10357   \n",
       "1    Mtb-m-macr-map-Mtb-ma-macr           15720           7           18217   \n",
       "2  Mtb-m-macr-map-COVID-ma-macr           15720           7            1630   \n",
       "3  Mtb-m-macr-map-merge-ma-macr           15720           7           19847   \n",
       "\n",
       "   _batch_que  \n",
       "0           3  \n",
       "1          41  \n",
       "2           3  \n",
       "3          44  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease = 'Mtb'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = df_apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "df_para_disease_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dbb00256-ceb8-48b7-b63a-a726faeab1b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] Mtb\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[start]240506-1732 Mtb_m-corss-h;csMAHN;Mtb-m-macr-map-Mtb-h-macr;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "                 Mtb_m    Mtb_h\n",
      "macr_cell_type                 \n",
      "AM_1            3218.0      NaN\n",
      "IM_3            2861.0      NaN\n",
      "AM_2            2547.0      NaN\n",
      "IM_2            1896.0      NaN\n",
      "IM_1            1852.0      NaN\n",
      "AM_3            1843.0      NaN\n",
      "AM_4             967.0      NaN\n",
      "IM_4             536.0      NaN\n",
      "macrophage         NaN  10357.0\n",
      "Task: refernece:Mtb_m 15720 cells x 22626 gene -> query:Mtb_h 10357 cells x 20631 gene in Mtb\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 28.472209215164185\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d67ab75-48f0-4330-b9f4-0a2fc56e3eaf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[start]240506-1734 Mtb_m-corss-h;csMAHN;Mtb-m-macr-map-Mtb-h-macr;epochs=500;is_1v1=False;240506-1734\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for _ in range(3):\n",
    "    n_epochs = [100, 200, 200]\n",
    "    for _i, _row in df_para_disease[df_para_disease['name_que'].isin('Mtb-h-macr,merge-ma-macr'.split(','))].iterrows():\n",
    "        run_cross_species_models(\n",
    "            path_adata1=_row['path_ref'],\n",
    "            path_adata2=_row['path_que'],\n",
    "            key_class1=_row['key_cell_type'],\n",
    "            key_class2=_row['key_cell_type'],\n",
    "            sp1=_row['sp_simple_ref'],\n",
    "            sp2=_row['sp_simple_que'],\n",
    "            tissue_name=_row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={};{}'.format(\n",
    "                _row['name_ref'],\n",
    "                _row['name_que'],\n",
    "                sum(n_epochs),\n",
    "                is_1v1,time_tag_get()),\n",
    "            resdir=p_res,\n",
    "            models='csMAHN'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b64345ea-c16f-43d3-8640-8a6967bfcecd",
   "metadata": {},
   "source": [
    "# LN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49f5557d-840f-45b2-9125-b1f0d66eeb9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease = 'LN'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = df_apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "df_para_disease_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fc42d6a-82d6-4a15-960f-0f7a9e6eaa9c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    # continue\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        # models='csMAHN,came'.split(','),\n",
    "        models='csMAHN'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# # SAMap\n",
    "# for _i, _row in df_para_disease.iterrows():\n",
    "#     run_cross_species_models(\n",
    "#         path_adata1=_row['path_ref'],\n",
    "#         path_adata2=_row['path_que'],\n",
    "#         key_class1=_row['key_cell_type'],\n",
    "#         key_class2=_row['key_cell_type'],\n",
    "#         sp1=_row['sp_simple_ref'],\n",
    "#         sp2=_row['sp_simple_que'],\n",
    "#         tissue_name=_row['tissue'],\n",
    "#         aligned=aligned,\n",
    "#         resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "#             _row['name_ref'],\n",
    "#             _row['name_que'],\n",
    "#             is_1v1),\n",
    "#         resdir=p_res,\n",
    "#         models='SAMap'.split(','),q\n",
    "#         **{\n",
    "#            'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d377dfe-78bc-4080-9d50-5504d320e4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n[finish]\\n\".center(100, \"-\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "publish",
   "language": "python",
   "name": "publish"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
